tcpproxy.go

```diff
@@ -122,25 +131,43 @@ func (p *Proxy) configFor(ipPort string) *config {
 		p.configs = make(map[string]*config)
 	}
 	if p.configs[ipPort] == nil {
-		p.configs[ipPort] = &config{}
+		p.configs[ipPort] = NewConfig()
 	}
 	return p.configs[ipPort]
 }
 
-func (p *Proxy) addRoute(ipPort string, r route) {
+func (p *Proxy) addRoute(ipPort string, r route) (routeId int) {
 	cfg := p.configFor(ipPort)
-	cfg.routes = append(cfg.routes, r)
+	cfg.Lock()
+	defer cfg.Unlock()
+	routeId = cfg.nextRouteId
+	cfg.nextRouteId++
+	cfg.routes[routeId] = r
+	return
 }
 
 // AddRoute appends an always-matching route to the ipPort listener,
-// directing any connection to dest.
+// directing any connection to dest. The added route's id is returned
+// for future removal.
 //
 // This is generally used as either the only rule (for simple TCP
 // proxies), or as the final fallback rule for an ipPort.
 //
 // The ipPort is any valid net.Listen TCP address.
-func (p *Proxy) AddRoute(ipPort string, dest Target) {
-	p.addRoute(ipPort, fixedTarget{dest})
+func (p *Proxy) AddRoute(ipPort string, dest Target) (routeId int) {
+	return p.addRoute(ipPort, fixedTarget{dest})
+}
+
+// RemoveRoute removes an existing route for ipPort. If the route is
+// not found, this is an no-op.
+//
+// Both AddRoute and RemoveRoute is go-routine safe.
+func (p *Proxy) RemoveRoute(ipPort string, routeId int) (err error) {
```

Comment:
un-name this, per https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

Reply:
Done. Thanks.